drivers/client_access_control/app/controllers/client_access_control/clients_controller.rb

```diff
@@ -76,6 +76,23 @@ def index
     end
   end
 
+  # FIXME: optimization and refactor needed, this is hack
+  # It seems that searchable_to scope is not-performant under some configurations.
+  protected def temporary_acl_search_hack
+    upper_bound_limit = 500 # protect from overly broad queries
+    # 1. Perform text search first
+    unauthorized_client_ids = client_source.text_search(params[:q], sorted: sorted).limit(upper_bound_limit).pluck(:id)
```

Comment:
I see, I'd missed that the first query wasn't bounded by the user's permission.  That makes more sense, and I agree, this should probably cast a wider net, though if your search matches 500, you probably need to refine it.

Maybe we should leave it big (5,000?).  

Maybe the first query should be 5,000 and the second could be limited to 500.  Maybe we just go with 500 as you have it written and see if there are issues.